### DISCUSSION TITLE
How to get available screen resolutions?

### DISCUSSION THREAD

**Discussion by XeonG** [AvaloniaUI/Avalonia#13737] [General]
that is all

**Comment by thevortexcloud** [comment]
Do you want the list of possible resolutions or the current resolutions? The latter can be done with the `Screen` class. EG `this.Screens.Primary` in a Window will return information about the current primary screen size. If you want the former option then as far as I know Avalonia does not have a built in method for it. If you are on Linux you can easily read the list from `/sys/class/drm/CARDNUMBER/DISPLAYNUMBER/modes` where CARDNUMBER is the number for your GPU and DISPLAYNUMBER is the display connected to that GPU. If you need a cross platform solution and don't mind bringing in another library, SDL2 can do this for you fairly easily.

https://wiki.libsdl.org/SDL2/SDL_DisplayMode

There is also this pretty good C# binding for SDL2

https://github.com/flibitijibibo/SDL2-CS

**Comment by XeonG** [comment]
I need a list of all possible resolutions supported by the display..  in the case of multiple monitors would expect some way to get the number of monitors and list of supported resolutions etc but also the current active screen resolution... seems abit silly a crossplatform framework made it a v11 and still doesn't have anything for this.

I might just use JavaFX, give up on C# Net stuff it's all just incomplete from just about everything from docs to api. In other news it seems AI had more faith in this than I do lol..


"Could you write a script to get the available screen resolutions that would work in c# Avalonia project?"

using System;
using System.Linq;
using Avalonia;
using Avalonia.Media;
using Avalonia.Platform;

public class ScreenResolutionHelper
{
    public static void Main()
    {
        var screens = Screen.AllScreens.OrderByDescending(screen => screen.WorkingArea.Width * screen.WorkingArea.Height);
        foreach (var screen in screens)
        {
            Console.WriteLine($"Screen Resolution: {screen.Bounds.Width}x{screen.Bounds.Height}, Name: {screen.DeviceName}");
        }
    }
}


yeaaaah

**Reply by thevortexcloud** [reply]
> seems abit silly a crossplatform framework made it a v11

The previous version was actually `.10`. It was decided to move it to `11` so people would not assume it was in beta or something. Avalonia in it's current state is perfectly usable. Querying all the available display resolutions however is not a use case most people need as such there are not a lot of people writing APIs to do it. As I mentioned there are ways to do it yourself fairly easily via native APIs. If you don't feel like using SDL you can also just do the interop yourself. EG on Windows you would need to do [this](https://stackoverflow.com/questions/744541/how-to-list-available-video-modes-using-c). In any case, the Avalonia `Screen` class is actually based on Winforms `Screen` class. Which also does not give all the available resolutions. You could probably open an issue to request that it be possible to do this.

https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.screen?view=windowsdesktop-8.0

>  C# Net stuff it's all just incomplete from just about everything from docs

I agree Avalonia's docs could be better. But a fairly large amount (probably ~60-70%) of Avalonia is interchangeable with WPF which does have much better documentation. Honestly I just google for WPF examples whenever I can't find an Avalonia one. It works most of the time. A large part of Avalonia's code has been either taken directly or has been ported from WPF, UWP and other Microsoft frameworks.

> I might just use JavaFX, give up on C# Net stuff it's all just incomplete from just about everything from docs to api. In other news it seems AI had more faith in this than I do lol..

I am fairly certain the code it generated wont even compile. `Screen` has no public static members in Avalonia. It does however have public static members on the Winforms version. Not to mention what it is doing is not what you asked for. I swear most AI code generation is actually useless...

**Reply by maxkatz6** [reply]
> I need a list of all possible resolutions supported by the display

Why? A typical app doesn't need to know about ALL possible resolutions for the display. Especially when it can be dynamic or unconstrained (any combination of width and height). Even more UI framework doesn't have to provide this information, as it's way too specific. There is nothing silly about it.

**Reply by maxkatz6** [reply]
> yeaaaah

```csharp
var window = new Window();
var screens = window.Screens.All;
```

And you have to do it in an Avaloina application. Not in the Main method. 
Before complaining about documentation, please read it first. Or at least start from the provided templates so you can have a buildable app.

**Comment by XeonG** [comment]
It's for a making a small desktop cross platform portable game launcher ui that needs to know what available resolutions are supported by the available monitors on that computer, so that it can provide those options to the user to choose and provide those as args to the game before launching...  but I'll just build it with javafx instead, they've got a screen builder ui that does actually support getting this information, I believe javfx is also a UI framework.. also Winforms was aswel and you could get that information.. things in MS land really have gotten shit.

> "Why? A typical app doesn't need to know about ALL possible resolutions for the display"

.... yeah sure... well I guess other developers might also want to not bother with a framework who's developers don't see the need for it support things that don't cross there mind for what an app should be able to do,  it begs the question of just what else is one going to find out later that it doesn't support.

**Reply by thevortexcloud** [reply]
>also Winforms was aswel and you could get that informatio

As I said, Winforms does not give you this information out of the box, You can get it. But you need to write code to do it yourself. Which as I also said is fairly straightforwards to do on most platforms. 

Video games are really the only scenario where I would think someone needs this information. Which is frankly not what most people are using a desktop UI framework for. Which is not to say there is not a valid use case for knowing this information in a desktop UI framework. 

I also just checked. From what I can tell JavaFX does not seem to give this information either. At best it just gives the current resolution of all screens. Just like Avalonia does (in fact [their Screen class ](https://docs.oracle.com/javafx/2/api/javafx/stage/Screen.html) is nearly identical to the Winforms and Avalonia one). I can't find any reference to this "screen builder" you are referring to. I can however find many references to something called "scene builder", which is apparently just a UI designer. So I am not sure what that has to do with it here. But I may be missing something.

AWT on the other does allow for this. But AWT has nothing directly to do with JavaFX.

https://docs.oracle.com/javase/8/docs/api/java/awt/GraphicsDevice.html

**Reply by cyraid** [reply]
Why is Screens in Window? What if I want to know the screens without a Window Instance? Like if I want to iterate the Screens and then make a new Window at the position of a specific Screen?

**Reply by thevortexcloud** [reply]
My comment is now out of date. There was a big refactor awhile back to make screens better.

https://github.com/AvaloniaUI/Avalonia/pull/16295/

Basically you just need some sort of top level now.

**Reply by cyraid** [reply]
@thevortexcloud I get it .. but why would it be a part of a top level to get all the screens? That means I have to make a window first (or view), then get the screens, then resize the window.

I don't know it this is cause avalonia gets a connection to the display only with a top level?

Would be nice to be able to get the screens available first before determining how big to make the window.

**Reply by maxkatz6** [reply]
You need to create a window, yes, but you don't need to actually show it before you read screens and position this window.

Screens being part of the TopLevel comes from platforms, where you actually need environment context first. Like android and iOS, but also Linux in multi-seat setup (single PC/OS, multiple desktops, each with their own set of displays). Although, Avalonia doesn't support multi-seat linux properly at the moment.

**Reply by cyraid** [reply]
Is it because of x11 display connection that gets connected on window create for example? (I'm just guessing, not sure if that's the point it does). In every circumstance that I can think of, like even with activity context, I'm sure you can get screen information before creating a window.

Like, create and get that connection pointer on application start? Or is it because of not having to have a connection if you don't need it.

Does multi seat need multiple display connections? And multiple screens per display connection? If so, why would you need a window first anyway?

Not intending to whine too much on it, just a little backwards thinking for me, like frying some eggs and then putting butter on it (instead of on the pan first).. Weird analogy but I'm writing while tired so go easy on me.

**Comment by XeonG** [comment]
I tried the updated one, it shows 'Frequency' in the debug for each screen, yet no way to get that in the code?
